### PR TITLE
[68_8] extract keywords from abstract-data and add them to exported pdf

### PR DESCRIPTION
**Pull request by jingkaimori**
## What work have you done in the current Pull Request?

- [x] add a new kind of metadata `keyword`, and corresponding searcher.
- [x] export this metadata to PDF document.
- [ ] related tests for all metadata entries, maybe in separate task

## How to test?

Insert abstraction into a document, and add keywords with focus toolbar. Then export this document to PDF, open exported PDF in browser, and open "document properties" dialogue.

### CODE REVIEW

File: src/Data/Convert/Texmacs/fromtm.cpp [hunk 1]
```diff
@@ -602,5 +602,25 @@ search_metadata (tree doc, string kind) {
     tree t= search_tag_quick (doc, "tmdoc-title");
     if (t != "") return tree_to_verbatim (t[0], false, "cork");
   }
+  tree ad= search_tag_quick (doc, "abstract-data");
```
Comment (darcy-shen) [inline src/Data/Convert/Texmacs/fromtm.cpp:605]: `ad` is not a good variable name, just use `abstract_data`

File: src/Data/Convert/Texmacs/fromtm.cpp [hunk 2]
```diff
@@ -602,5 +602,25 @@ search_metadata (tree doc, string kind) {
...
+  if (ad != "") {
```
Comment (darcy-shen) [inline src/Data/Convert/Texmacs/fromtm.cpp:606]: `!is_empty (abstract_data)` would be better

File: src/Data/Convert/Texmacs/fromtm.cpp [hunk 3]
```diff
@@ -602,5 +602,25 @@ search_metadata (tree doc, string kind) {
...
+            string r= tree_to_verbatim (ak[0], false, "cork");
+            for (int i= 1; i < N (ak); i++) {
+              r << ", " << tree_to_verbatim (ak[i], false, "cork");
+            }
+            return r;
+          }
+        }
+      }
+    }
+  }
```
Comment (darcy-shen) [inline src/Data/Convert/Texmacs/fromtm.cpp:624]: unit tests should be added for search_metadata

File: src/Data/Convert/Texmacs/fromtm.cpp (hunk 1 above)
Comment (darcy-shen) [reply src/Data/Convert/Texmacs/fromtm.cpp:605]: `dd` is also not a good variable name, should use `doc_data` instead

Comment (jingkaimori) [reply src/Data/Convert/Texmacs/fromtm.cpp:605]: what about `ak`?

File: src/Data/Convert/Texmacs/fromtm.cpp [hunk 4]
```diff
@@ -591,16 +591,36 @@ search_metadata_tag (tree doc, string tag) {
 
 string
 search_metadata (tree doc, string kind) {
-  tree dd= search_tag_quick (doc, "doc-data");
-  if (dd != "") {
+  tree doc_data= search_tag_quick (doc, "doc-data");
+  if (doc_data != "") {
```
Comment (darcy-shen) [inline src/Data/Convert/Texmacs/fromtm.cpp:595]: doc_data is a tree, `doc_data != ""` is not a good practice. There are implicit conversion. I guess `N(doc_data) > 0` is better

File: src/Data/Convert/Texmacs/fromtm.cpp [hunk 5]
```diff
@@ -591,16 +591,36 @@ search_metadata_tag (tree doc, string tag) {
...
     if (kind == "author")
-      return search_metadata_tag (dd, "author-name");
+      return search_metadata_tag (doc_data, "author-name");
   }
   if (kind == "title") {
     tree t= search_tag_quick (doc, "tmdoc-title");
     if (t != "") return tree_to_verbatim (t[0], false, "cork");
   }
+  tree abstract_data= search_tag_quick (doc, "abstract-data");
+  if (!is_empty (abstract_data)) {
```
Comment (darcy-shen) [inline src/Data/Convert/Texmacs/fromtm.cpp:606]: same as doc_data

Comment (darcy-shen) [reply src/Data/Convert/Texmacs/fromtm.cpp:606]: I thought abstract_data is string, `is_empty` is for string and I remember it is not for tree

`is_empty` is identical to `N(xxx) == 0`, and I think it is better than `N(xxx) == 0`, for tree, we may need to add the is_empty inline routine.

File: src/Data/Convert/Texmacs/fromtm.cpp [hunk 6]
```diff
@@ -591,16 +591,36 @@ search_metadata_tag (tree doc, string tag) {
 
 string
 search_metadata (tree doc, string kind) {
-  tree dd= search_tag_quick (doc, "doc-data");
```
Comment (darcy-shen) [inline src/Data/Convert/Texmacs/fromtm.cpp:594]: the branching in this routine is a bit messed up

the main condition should be kind:

if (kind == "title") {
} else if (kind == "author") {
} else if (kind == "keyword") {
} else {
  error handling on invalid value
}

File: src/Data/Convert/Texmacs/fromtm.cpp (hunk 4 above)
Comment (jingkaimori) [reply src/Data/Convert/Texmacs/fromtm.cpp:595]: `search_tag_quick` may return atomic tree with empty string, on which `N(t)` procedure will crash. `is_compound(doc_data)` may be better choice.

File: tests/Data/Convert/convert_test.cpp [hunk 7]
```diff
@@ -0,0 +1,69 @@
+/******************************************************************************
+ * MODULE     : converter_test.cpp
+ * DESCRIPTION: Properties of characters and strings
+ * COPYRIGHT  : (C) 2019 Darcy Shen
```
Comment (darcy-shen) [inline tests/Data/Convert/convert_test.cpp:4]: should be your own name

File: tests/Data/Convert/convert_test.cpp [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+                                   "keyword 2", "keyword 3")))
+      << string ("Test of Metadata") << string ("author1, author2, author3")
+      << string ("keyword 1, keyword 2, keyword 3");
+  QTest::newRow ("texmacs document")
+      << tree (DOCUMENT, compound ("tmdoc-title", "Test of manual title"))
+      << string ("Test of manual title") << string ("") << string ("");
+}
+
+void
+TestConverter::test_search_metadata () {
```
Comment (darcy-shen) [inline tests/Data/Convert/convert_test.cpp:58]: what if the kind is invalid

File: src/Data/Convert/Texmacs/fromtm.cpp (hunk 1 above)
Comment (darcy-shen) [reply src/Data/Convert/Texmacs/fromtm.cpp]: just use a long name

**Pull request merged**